inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/hook/HookManager.java

```diff
@@ -0,0 +1,221 @@
+package rocks.inspectit.oce.core.instrumentation.hook;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import io.opencensus.stats.Aggregation;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import net.bytebuddy.description.method.MethodDescription;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.event.EventListener;
+import org.springframework.stereotype.Service;
+import rocks.inspectit.oce.bootstrap.Instances;
+import rocks.inspectit.oce.bootstrap.instrumentation.IMethodHook;
+import rocks.inspectit.oce.bootstrap.instrumentation.noop.NoopHookManager;
+import rocks.inspectit.oce.bootstrap.instrumentation.noop.NoopMethodHook;
+import rocks.inspectit.oce.core.config.InspectitConfigChangedEvent;
+import rocks.inspectit.oce.core.config.InspectitEnvironment;
+import rocks.inspectit.oce.core.config.model.instrumentation.InternalSettings;
+import rocks.inspectit.oce.core.instrumentation.config.InstrumentationConfigurationResolver;
+import rocks.inspectit.oce.core.instrumentation.config.event.InstrumentationConfigurationChangedEvent;
+import rocks.inspectit.oce.core.instrumentation.config.model.MethodHookConfiguration;
+import rocks.inspectit.oce.core.instrumentation.event.IClassDiscoveryListener;
+import rocks.inspectit.oce.core.selfmonitoring.SelfMonitoringService;
+import rocks.inspectit.oce.core.service.BatchJobExecutorService;
+import rocks.inspectit.oce.core.utils.CommonUtils;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import java.lang.instrument.Instrumentation;
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Service
+public class HookManager implements IClassDiscoveryListener {
```

Comment:
Java docs on the methods and some small inline explanations would really be helpful here..